
# 1) CASE COGNITIVO



OLA, esse é o trabalho do Felipe H B Bittar para o case da COGNITIVO.AI
irei explicar através dos textos o que foi feito em cada pedaço de código.
Lembre-se, o código é algo ''pessoal'' e, as vezes, leva tempo para entender o que a pessoa fez. Entretanto, um código bem comentado auxilia ao leitor a entender de maneira MACRO o que está sendo feito. Assim, com a geração deste case espero demonstrar habilidades mínimas em PYTHON. JUPYTER NOTEBOOK, DATA CLEANING E DATA TRANSFORMATION DE DATABASES 'CONFUSOS' RETIRADOS DA INTERNET, FEATURE SELECTION, ML e METODOLOGIA CIENTIFICA NA ESCOLHA DE MODELOS.

**O objetivo deste teste é**: Criar um modelo de PRICING de estadias/alugueis. Esse o algoritmo de ML utilizado será as RNA's e um teste de variação de topologias (adição de neuronios da na camada escondida) foi feito visando, por meio de testes empiricos, qual foi o melhor desempenho entre as redes.

No código abaixo foi feito:

*   Importação das bibliotecas
*   Carregamento do arquivo .csv que está no meu github
*   O arquivo .CSV já foi previamente  manualmente e, posteriormente, digitalmente 'limpado' , retirando-se as entradas/features que CERTAMENTE não tem nenhuma relevancia ou correlação com a saída(pricing) e/ou estavam tornando o modelo muito complexo/oneroso
*   Criação de um dataframe dos dados do arquivo .csv
* criação de nomeação das colunas/features em : Feature 1, Feature 2...Feature n
* Em caso de falta de dados ou NaN, foi adotado a politica de substituir este valor pela média.
*   Codificação numérica dos dados tipo string ( nome de bairro --> encoding). onde 1==Zona Sul , 2==centro , 3== zona oeste, 4==todo o resto





In [305]:
import pandas as pd
import numpy as np
import sys
from sklearn.preprocessing import OrdinalEncoder
import matplotlib.pyplot as plt
from sklearn.neural_network import MLPRegressor
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn import metrics
import warnings
warnings.filterwarnings('ignore')



np.set_printoptions(threshold=sys.maxsize)
url='https://raw.githubusercontent.com/felipehugobragabittar/Cognitivo/main/listings.csv'

df = pd.read_csv(url)
df = df.iloc[: , :10]
num_entradas=len(df.index)
num_colunas = len(df.columns)
#print(num_entradas)
#print(num_colunas)
df.columns=["Feature"+str(i) for i in range(1, num_colunas+1)]
df=df.loc[:, ~df.columns.isin(['Feature3','Feature4'])]


In [306]:
#print(df.dtypes)
#df.info()
df1 = df.iloc[:,0]        #peguei a primeira coluna
u=df1.unique()
#print(u)

aux=df.index[df['Feature1']=='Vila Isabel'].tolist()
df = df.drop(aux)
aux=df.index[df['Feature1']=='https://a0.muscache.com/im/pictures/user/438b48fc-37b3-446e-9704-1f98df288291.jpg?aki_policy=profile_x_medium'].tolist()
df = df.drop(aux)
aux=df.index[df['Feature1']=='nan'].tolist()
df = df.drop(aux)
aux=df.index[df['Feature1']=='Barra da Tijuca'].tolist()
df = df.drop(aux)
aux=df.index[df['Feature1']=='Tijuca'].tolist()
df = df.drop(aux)
aux=df.index[df['Feature1']=='Santa Teresa'].tolist()
df = df.drop(aux)

df['Feature1'] = df['Feature1'].str.replace(',','')

df1 = df.iloc[:,0] 
u=df1.unique()
#print(u)
df['Feature1'] = df['Feature1'].astype(float)
df['Feature1'].fillna(value=df['Feature1'].mean(), inplace=True)

df1 = df.iloc[:,0] 
u=df1.unique()
#print(u)

df['Feature1'].isnull().sum()
df['Feature1'].isnull().values.any()



False

In [307]:
df1 = df.iloc[:,1]        #peguei a primeira coluna
u=df1.unique()
#print(u)

aux=df.index[df['Feature2']=='within a few hours'].tolist()
df = df.drop(aux)
aux=df.index[df['Feature2']=='f'].tolist()
df = df.drop(aux)
df['Feature2'] = df['Feature2'].str.replace(',','.')
df['Feature2'] = df['Feature2'].astype(float)
df['Feature2'].fillna(value=df['Feature2'].mean(), inplace=True)

df1 = df.iloc[:,1]        #peguei a primeira coluna
u=df1.unique()
#print(u)

df['Feature2'].isnull().sum()
df['Feature2'].isnull().values.any()


False

In [308]:
df1 = df.iloc[:,2]        #peguei a primeira coluna
u=df1.unique()
#print(u)

aux=df.index[df['Feature5']=='https://a0.muscache.com/im/pictures/user/f07b43c8-5895-43a6-9d09-a54ce8ae022b.jpg?aki_policy=profile_x_medium'].tolist()
df = df.drop(aux)
aux=df.index[df['Feature5']=='t'].tolist()
df = df.drop(aux)
aux=df.index[df['Feature5']=='1,00'].tolist()
df = df.drop(aux)

num_entradas=len(df.index)


for i in range(0,num_entradas):
  if df['Feature5'].iloc[i] in ['Copacabana','Santa Teresa','Gávea','Leblon','Ipanema','Laranjeiras','Urca','Botafogo','Leme','Jardim Botânico','Lagoa',
  'Humaitá','Flamengo','Flamengo, Rio de Janeiro, Brazil','Copacabana, Rio de Janeiro, Brazil','Ipanema, Rio de Janeiro, Brazil',
  'Glória, Rio de Janeiro, Brazil']:
    df['Feature5'].iloc[i]=1

  elif df['Feature5'].iloc[i] in ['Centro','Santo Cristo','Estácio','Catete','Cosme Velho','Santa Teresa','Cidade Nova','Glória','São Cristóvão','Tijuca']:
    df['Feature5'].iloc[i]=2

  elif df['Feature5'].iloc[i] in ['Jacarepaguá','Barra da Tijuca','Recreio dos Bandeirantes','Joá'
 'Alto da Boa Vista','São Conrado' ]:
    df['Feature5'].iloc[i]=3
  else:
    df['Feature5'].iloc[i]=0



df1 = df.iloc[:,2]        #peguei a primeira coluna
u=df1.unique()
#print(u)
df['Feature5'] = df['Feature5'].astype(float)

#df['Feature5'].isnull().sum()
#df['Feature5'].isnull().values.any()


[1 3 2 0]


In [309]:
df1 = df.iloc[:,3]        #peguei a primeira coluna
u=df1.unique()
#print(u)

aux=df.index[df['Feature6']=='Entire home/apt'].tolist()
df = df.drop(aux)
aux=df.index[df['Feature6']=='Private room'].tolist()
df = df.drop(aux)
df['Feature6'] = df['Feature6'].str.replace(',','.')
df['Feature6'] = df['Feature6'].astype(float)


df1 = df.iloc[:,3]        #peguei a primeira coluna
u=df1.unique()
#print(u)

df['Feature6'].isnull().sum()
df['Feature6'].isnull().values.any()

False

In [310]:
df1 = df.iloc[:,4]        #peguei a primeira coluna
u=df1.unique()
#print(u)

aux=df.index[df['Feature7']==' half-'].tolist()
df = df.drop(aux)
aux=df.index[df['Feature7']=='Half-'].tolist()
df = df.drop(aux)
df['Feature7'] = df['Feature7'].str.replace(',','.')
df['Feature7'] = df['Feature7'].astype(float)
df['Feature7'].fillna(value=df['Feature7'].mean(), inplace=True)


df1 = df.iloc[:,4]        #peguei a primeira coluna
u=df1.unique()
#print(u)
df['Feature7'].isnull().sum()
df['Feature7'].isnull().values.any()


Feature1  Feature2  Feature5  Feature6  Feature7 Feature8 Feature9  \
0           0.0       0.0       1.0      12.0       4.0     4,00     8,00   
1         700.0       7.0       1.0      16.0       5.0     7,00    10,00   
2         800.0       8.0       1.0       4.0       1.0     1,00     2,00   
3         200.0       2.0       1.0       2.0       2.0     1,00     1,00   
4         700.0       7.0       1.0       5.0       1.0     1,00     3,00   
...         ...       ...       ...       ...       ...      ...      ...   
24478       0.0       0.0       1.0      10.0       3.0     4,00     8,00   
24479     100.0       1.0       1.0       4.0       1.0     1,00     2,00   
24480     100.0       1.0       1.0       8.0       2.0     3,00     4,00   
24481     900.0       9.0       1.0       6.0       1.0     1,00     8,00   
24482    6500.0      65.0       3.0       3.0       1.0     1,00     2,00   

       Feature10  
0      $3,149.00  
1        $225.00  
2        $373.00  
3        $243.00  
4        $690.00  
...          ...  
24478  $1,786.00  
24479    $460.00  
24480    $891.00  
24481    $177.00  
24482    $818.00  

[24439 rows x 8 columns]

In [311]:
df1 = df.iloc[:,5]        #peguei a primeira coluna
u=df1.unique()
#print(u)


df['Feature8'] = df['Feature8'].str.replace(',','.')
df['Feature8'] = df['Feature8'].astype(float)
df['Feature8'].fillna(value=df['Feature8'].mean(), inplace=True)

df['Feature8'].isnull().sum()
df['Feature8'].isnull().values.any()

False

In [312]:
df1 = df.iloc[:,6]        #peguei a primeira coluna
u=df1.unique()
#print(u)


df['Feature9'] = df['Feature9'].str.replace(',','.')
df['Feature9'] = df['Feature9'].astype(float)
df['Feature9'].fillna(value=df['Feature9'].mean(), inplace=True)

df['Feature9'].isnull().sum()
df['Feature9'].isnull().values.any()

False

In [313]:
df1 = df.iloc[:,7]        #peguei a primeira coluna
u=df1.unique()
#print(u)

df['Feature10'] = df['Feature10'].str.replace('$','')
df['Feature10'] = df['Feature10'].str.replace(',','')
df['Feature10'] = df['Feature10'].astype(float)


df1 = df.iloc[:,7]        #peguei a primeira coluna
u=df1.unique()
#print(u)

df['Feature10'].isnull().sum()
df['Feature10'].isnull().values.any()

False

In [314]:
num_entradas=len(df.index)
num_colunas = len(df.columns)
#print(num_entradas)
#print(num_colunas)
df.columns=["Feature"+str(i) for i in range(1, num_colunas+1)]

# 2) CASE COGNITIVO
# **Disposição do Dataframe resultante:**

Database final composto por 7 entradas e 1 saida
possui 24439 amostras para o treinamento: 

In [315]:
df

Feature1  Feature2  Feature3  Feature4  Feature5  Feature6  Feature7  \
0           0.0       0.0       1.0      12.0       4.0       4.0       8.0   
1         700.0       7.0       1.0      16.0       5.0       7.0      10.0   
2         800.0       8.0       1.0       4.0       1.0       1.0       2.0   
3         200.0       2.0       1.0       2.0       2.0       1.0       1.0   
4         700.0       7.0       1.0       5.0       1.0       1.0       3.0   
...         ...       ...       ...       ...       ...       ...       ...   
24478       0.0       0.0       1.0      10.0       3.0       4.0       8.0   
24479     100.0       1.0       1.0       4.0       1.0       1.0       2.0   
24480     100.0       1.0       1.0       8.0       2.0       3.0       4.0   
24481     900.0       9.0       1.0       6.0       1.0       1.0       8.0   
24482    6500.0      65.0       3.0       3.0       1.0       1.0       2.0   

       Feature8  
0        3149.0  
1         225.0  
2         373.0  
3         243.0  
4         690.0  
...         ...  
24478    1786.0  
24479     460.0  
24480     891.0  
24481     177.0  
24482     818.0  

[24439 rows x 8 columns]

# 3) CASE COGNITIVO

Nesta segunda etapa, Será feito um teste de '' seleção de features''(Feature Selection) , por meio da correlação de Pearson com o objetivo de saber quais as entradas tem mais correlação com a saida.

No código abaixo foi feito:

*   Estudo da correlação entradas vs saida 
*  plot da matrix de correlação
*   Foi observada uma redundância das Features 1 e 2, onde a Feature 2 apresentam os mesmos valores divididos por 100 da Feature 1.
*   Remoção da feature 2 e renomeação do cabeçalho do dataframe




In [316]:
corr = df.corr()

corr.style.background_gradient(cmap='coolwarm')


# **Disposição do Dataframe resultante:**

Database final composto por 6 entradas e 1 saida
possui 24439 amostras para o treinamento: 

In [317]:
df=df.loc[:, ~df.columns.isin(['Feature2'])]

num_colunas = len(df.columns)
#print(num_colunas)
df.columns=["Feature"+str(i) for i in range(1, num_colunas+1)]
df

Feature1  Feature2  Feature3  Feature4  Feature5  Feature6  Feature7
0           0.0       1.0      12.0       4.0       4.0       8.0    3149.0
1         700.0       1.0      16.0       5.0       7.0      10.0     225.0
2         800.0       1.0       4.0       1.0       1.0       2.0     373.0
3         200.0       1.0       2.0       2.0       1.0       1.0     243.0
4         700.0       1.0       5.0       1.0       1.0       3.0     690.0
...         ...       ...       ...       ...       ...       ...       ...
24478       0.0       1.0      10.0       3.0       4.0       8.0    1786.0
24479     100.0       1.0       4.0       1.0       1.0       2.0     460.0
24480     100.0       1.0       8.0       2.0       3.0       4.0     891.0
24481     900.0       1.0       6.0       1.0       1.0       8.0     177.0
24482    6500.0       3.0       3.0       1.0       1.0       2.0     818.0

[24439 rows x 7 columns]

# 4) CASE COGNITIVO

No código abaixo foi feito:

*   Criação do dataset de treino e de teste com a regra 90-10% , onde 90% dos dados são usados para treino e 10% para teste de desempenho
*   Normalização das entradas/saidas para que não haja saturação da FUNÇÃO DE ATIVAÇÃO (tangente hiperbólica) DA RNA 


In [324]:
y= df['Feature7']
x=df.loc[:, ~df.columns.isin(['Feature7'])]
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.1)


norm = MinMaxScaler().fit(x_train)
x_train = norm.transform(x_train)
x_test = norm.transform(x_test)


print(len(x_train))
print(len(x_test))
print(len(y_train))
print(len(y_test))


ValueError: ignored



# 5) CASE COGNITIVO

No código abaixo foi feito:

*  Testes de TOPOLOGIA das RNAS
*   foram testadas as topologias com 10, 15 e 20 neuronios na primeira camada escondida (mantendo-se a taxa de aprendizado/otimizações).



In [328]:
clf1 = MLPRegressor(solver='adam', alpha=1e-5, activation='tanh', max_iter=1000, hidden_layer_sizes=(10,))
clf1.fit(x_train, y_train)
expected_y  = y_test
predicted_y = clf1.predict(x_test)
print(metrics.r2_score(expected_y, predicted_y))
print(metrics.mean_squared_log_error(expected_y, predicted_y))



clf2 = MLPRegressor(solver='adam', alpha=1e-5, activation='tanh', max_iter=1000, hidden_layer_sizes=(15,))
clf2.fit(x_train, y_train)
expected_y  = y_test
predicted_y = clf2.predict(x_test)
print(metrics.r2_score(expected_y, predicted_y))
print(metrics.mean_squared_log_error(expected_y, predicted_y))



clf3 = MLPRegressor(solver='adam', alpha=1e-5, activation='tanh', max_iter=1000, hidden_layer_sizes=(20,))
clf3.fit(x_train, y_train)
expected_y  = y_test
predicted_y = clf3.predict(x_test)
print(metrics.r2_score(expected_y, predicted_y))
print(metrics.mean_squared_log_error(expected_y, predicted_y))

0.019201905588815293
1.023099546340995
0.04576194378245313
0.9716912892599295
0.05975403797088097
0.9690913015114249


Através do teste empirico e pro reprodução por tréplica, o melhor modelo nos 3 testes cujo erro médio quadratico (MSE) apresentou o menor valor foi a topologia 6-20-1

MSE DOS MODELOS 1,2 e 3 respectivamente


**PRIMEIRO TESTE:**
modelo 1)  1.004782742051108
modelo 2)   0.9713643165977944
modelo 3)   0.959600611040562

**SEGUNDO TESTE:**
modelo 1)  1.0064875295710731
modelo 2)  0.9793350609991945
modelo 3)  0.9739452114455832

**TERCEIRO TESTE**:
modelo 1) 1.023099546340995
modelo 2) 0.9716912892599295
modelo 3) 0.9690913015114249